llmfoundry/eval/metrics/nlp.py

```diff
@@ -0,0 +1,583 @@
+# Copyright 2024 MosaicML LLM Foundry authors
+# SPDX-License-Identifier: Apache-2.0
+
+# Copyright 2022 MosaicML Composer authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""A collection of common torchmetrics for NLP tasks."""
+
+import logging
+import os
+import re
+import string
+import warnings
+from typing import Any, Dict, List
+from composer.utils import dist
+import numpy as np
+import torch
+from composer.utils.eval_client import (EvalClient, LambdaEvalClient,
+                                        LocalEvalClient,
+                                        MosaicMLLambdaEvalClient)
+from torch import Tensor
+from torch.nn import functional as F
+from torchmetrics import Metric
+
+log = logging.getLogger(__name__)
+
+__all__ = [
+    'InContextLearningMetric',
+    'InContextLearningLMAccuracy',
+    'InContextLearningMultipleChoiceAccuracy',
+    'InContextLearningGenerationWithAnswersTaskDataset',
+    'InContextLearningCodeEvalAccuracy',
+    'InContextLearningLMExpectedCalibrationError',
+    'InContextLearningMCExpectedCalibrationError',
+]
+
+
+class InContextLearningMetric(Metric):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.needs_batch = True
+
+    def update(
+        self,
+        batch: dict,
+        outputs: torch.Tensor,
+        labels: torch.Tensor,
+    ):
+        """Abstract interface for computing an in-context learning metrics.
+
+        The `output_logits` argument is deprecated and will be removed in v0.21 while it's functionality will
+        be moved to `outputs`.
+
+        Args:
+            batch (dict): Batch must consist minimally of `input_ids` as well as any other structure needed
+                to compute the metric.
+            output_logits (torch.Tensor): The model outputs evaluated on the batch `input_ids`
+            labels (torch.Tensor): The correct outputs.
+
+        Raises:
+            NotImplementedError: Abstract method must be implemented by subclasses
+        """
+        raise NotImplementedError
+
+
+class InContextLearningGenerationAccuracy(InContextLearningMetric):
```

Comment:
We check whether the generated output equals one of the answers, after applying some sort of post-processing to the generated output

Reply:
This includes regex searching, splitting up the result, normalizing, or anything else you could dream of

Reply:
wait my bad...the comments above are valid for the refactor I'm working on next ^^
I can change this to exact match accuracy since that's what this type of accuracy would be referred to as in common NLP parlance.